cur-lib/cur/stdlib/pattern-tree.rkt

```diff
@@ -0,0 +1,717 @@
+                                 ; based on proper inference
+                                 [extended-env (extend-env-with-tmps
+                                                env
+                                                (first (C-group-head-patterns merged-group))
+                                                tmp-map-with-ids
+                                                current-match-var)]
+                                 ; to ensure that we don't erroneously shadow any bindings, we need to generate temporaries for
+                                 ; inner arguments too; (s x) -> (s x1)
+                                 [fresh-result (generate-fresh-arguments (C-group-head-patterns merged-group) (C-group-bodies merged-group) tmp-map-with-ids current-match-var extended-env)]
+                                 [fresh-head-patterns (car fresh-result)]
+                                 [fresh-bodies (cdr fresh-result)]
+                                 ; for the match pattern, just fetch anything that's not a pattern variable
+                                 [match-pat (first fresh-head-patterns)]
+                                 ; hack: attach the temporary type based on the value we've assigned to it in the environment
+                                 ; TODO PR103: This can't possibly work... we're
+                                 ; trying to add a type to t by using t's type?
+                                 ; No, because the extended-env might contain
+                                 ; types for them computed from the pattern.
+                                 [tmp-map-with-ids-typed (map (lambda (t) (and t (syntax-property t ': (get-typeof t #:env extended-env))))
+                                                              tmp-map-with-ids)]
+                                 [new-match-vars (append (filter (compose not false?) tmp-map-with-ids-typed) (rest match-vars))])
+                            (create-pattern-tree-match-helper
+                             new-match-vars
+                             ; it may make our lives easier to keep track of the property on the match pattern itself
+                             (syntax-property match-pat 'is-pattern-variable? (C-group-is-pattern-variable? merged-group))
+                             (C-group-pattern-sub-matrix merged-group)
+                             fresh-bodies
+                             extended-env))))])
+      pt))
+
+  ;; For the current match variable and the head pattern for a particular
+  ;; pattern row, either creates a new C-group entry within the constructor
+  ;; hash or merges the remaining patterns to form a new pattern sub-matrix.
+  (define (process-pattern match-var head-pattern remaining-patterns body idx
+                           C-hash env)
+    ; TODO PR103: Not sure if this comment is relevant to the custom hash.
```

Comment:
Yeah, so in the case of efficiency it's only minor pruning at best. The hash acts more like a set of 2-tuples, where it's convenient to retrieve the value by a key.